File: Solution-Java/Permutation_567.java
public class Permutation_567 {
    // Refer to sliding window in Find_438
    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) return false;
        int[] freq = new int[256];
        for (char c : s1.toCharArray()) freq[c]++;
        int left = 0, right = 0, count = s1.length();
        while (right < s2.length()) {
            if (freq[s2.charAt(right++)]-- > 0) count--;
            if (count == 0) return true;
            if (right - left == s1.length() && ++freq[s2.charAt(left++)] > 0) count++;
        }
        return false;
    }
}

File: Solution-Python/Permutation_567.py
from collections import defaultdict

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        freq = defaultdict(int)
        for c in s1:
            freq[c] += 1
        l, r, count = 0, 0, len(s1)
        while r < len(s2):
            if freq[s2[r]] > 0:
                count -= 1
            freq[s2[r]] -= 1
            r += 1
            if count == 0:
                return True
            if r - l == len(s1):
                if freq[s2[l]] >= 0:
                    count += 1
                freq[s2[l]] += 1
                l += 1
        return False